### PR TITLE
Fix metrics reporting in kube-apiserver

### LINKED ISSUE #95556: Calls with verb=GET are reported with verb=LIST in metrics
**What happened**:
While I was watching grafana for a single run, I noticed that GET requests are reported as LIST requests, e.g. grafana shows ~1K QPS for LIST of pods
8vS47hgQ9GxtJhL
while kube-apiserver's logs show:
```
I1014 12:24:03.449859      11 httplog.go:89] "HTTP" verb="GET" URI="/api/v1/namespaces/test-mxsrii-31/pods/small-deployment-65-75785d8948-rwmdg" latency="5.177181ms" userAgent="kubelet/v1.20.0 (linux/amd64) kubernetes/ee297b6" srcIP="10.40.7.26:46666" resp=200
I1014 12:24:03.449819      11 httplog.go:89] "HTTP" verb="GET" URI="/api/v1/namespaces/test-mxsrii-17/pods/medium-deployment-0-698b4b86b9-fchjq" latency="6.536801ms" userAgent="kubelet/v1.20.0 (linux/amd64) kubernetes/ee297b6" srcIP="10.40.17.27:33852" resp=200
I1014 12:24:03.450079      11 httplog.go:89] "HTTP" verb="GET" URI="/api/v1/namespaces/test-mxsrii-33/pods/big-deployment-0-76677bd886-hthls" latency="7.076651ms" userAgent="kubelet/v1.20.0 (linux/amd64) kubernetes/ee297b6" srcIP="10.40.10.81:53078" resp=200
I1014 12:24:03.450385      11 httplog.go:89] "HTTP" verb="GET" URI="/api/v1/namespaces/test-mxsrii-31/pods/small-deployment-231-588fb44685-6ntjv" latency="7.255073ms" userAgent="kubelet/v1.20.0 (linux/amd64) kubernetes/ee297b6" srcIP="10.40.17.89:36454" resp=200
```
(there are no pod lists that time).

I'm not sure if this is intended change to metrics reporting (kind of unexpected to be honest), but this definitely affects way our scalability tests validate latency metrics:
* We do not validate GET API call latency at all
* Validation for LIST requests is also affected, as usually GET API calls are way cheaper than LIST requests, so grouping GET-LISTs and LIST-LISTs together lowers p99 for LISTs.

Also perf-dash shows that GET requests simply disappeared at some point:
http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadRequestCount_Prometheus&Resource=deployments&Scope=namespace&Subresource=&Verb=GET

image

**What you expected to happen**:
GET requests should be reported with verb=GET in apiserver_request_total metric.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix https://github.com/kubernetes/kubernetes/issues/95556

```release-note
Fix verb & scope reporting for kube-apiserver metrics (LIST reported instead of GET)
```

/sig apimachinery
/kind bug
/priority important-soon

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 1]
```diff
@@ -590,7 +590,7 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 			requestScope = "namespace"
 			namespaced = "Namespaced"
 		}
-		if strings.HasSuffix(action.Path, "/{path:*}") {
+		if strings.HasSuffix(action.Path, "/{name}") || strings.HasSuffix(action.Path, "/{path:*}") {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:593]: More context is in: https://github.com/kubernetes/kubernetes/issues/95556

Re https://github.com/kubernetes/kubernetes/issues/95556#issuecomment-708431219 after thinking about this, I think that if a request is touching only a single object, it should be reported as scope "request" (not as scope "namespace").

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:593]: Re adding a test - I think that adding a test is potentially non-trivial here, and given that we have broken reporting at this point after #93523 exposed this bug, I would like to fix the metrics to ensure that we start tracking performance the right way and add a test independently.

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:593]: I probably agree with the classification.

However I can't agree with leaving this untested, this is kind of critical to the functioning of the system. Is there really not an existing test for this?! (I can't find one if there is...) It looks like apiserver_test.go has a framework which should make it not too hard to set up a test server and send various URL paths to it?

@deads2k

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:593]: > However I can't agree with leaving this untested, this is kind of critical to the functioning of the system.

Agree

> Is there really not an existing test for this?! (I can't find one if there is...)

The fact that it's currently broken indicates there is not ;-)

> It looks like apiserver_test.go has a framework which should make it not too hard to set up a test server and send various URL paths to it?

We have TestApiserverMetrics that looks like it does the setup already. It seems like we can make more requests to a variety of endpoints and verify the resulting metrics include all the dimensions/label values we'd expect.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:593]: I didn't find any test too.

I don't see anything that is particularly useful in apiserver_test.

But I really like Jordan`s suggestion about the metrics test. Using this framework should actually be much less work.
I will try to get to it today.

[BTW - I'm glad that you agree with the fix at this point.]

**Comment by wojtek-t** [comment]
/hold
Until tests are added

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:593]: > I probably agree with the classification.

It seems reasonable to classify this way.  It's a little weird that it impacts the verb chosen though.

**Comment by wojtek-t** [comment]
I added the test.
/hold cancel

@deads2k @lavalamp - PTAL

File: test/integration/metrics/metrics_test.go [hunk 2]
```diff
@@ -111,3 +114,130 @@ func TestApiserverMetrics(t *testing.T) {
...
+				return false
+			}
+		}
+		return true
+	}
+
+	for _, expectedMetric := range expectedMetrics {
+		found := false
+		for _, sample := range samples {
+			if hasLabels(sample.Metric, expectedMetric) {
```
Comment (lavalamp) [inline test/integration/metrics/metrics_test.go:234]: I think this methodology won't detect if two requests have opposite bugs, i.e. some aspect of their metrics are flipped?

**Comment by lavalamp** [comment]
/lgtm
/approve

This is much better than no test, but can you file an issue or send a follow-up to improve the test?

File: test/integration/metrics/metrics_test.go (hunk 2 above)
Comment (wojtek-t) [reply test/integration/metrics/metrics_test.go:234]: Opened https://github.com/kubernetes/kubernetes/issues/95636

**Pull request merged**